evm/vm/forks/sharding/vm_state.py

```diff
@@ -185,3 +189,97 @@ def execute_transaction(self, transaction):
 
     def validate_transaction(self, transaction):
         validate_sharding_transaction(self, transaction)
+
+    def add_transaction(self, transaction, computation, block):
+        """
+        Add a transaction to the given block and
+        return `trie_data` to store the transaction data in chaindb in VM layer.
+
+        Update the bloom_filter, transaction trie and receipt trie roots, bloom_filter,
+        bloom, and used_gas of the block.
+
+        :param transaction: the executed transaction
+        :param computation: the Computation object with executed result
+        :param block: the Block which the transaction is added in
+        :type transaction: Transaction
+        :type computation: Computation
+        :type block: Block
+
+        :return: the block and the trie_data
+        :rtype: (Block, dict[bytes, bytes])
+        """
+        receipt = self.make_receipt(transaction, computation)
+        self.add_receipt(receipt)
+
+        # Create a new Block object
+        block_header = block.header.clone()
+        transactions = list(block.transactions)
+        block = self.block_class(block_header, transactions)
+
+        block.transactions.append(transaction)
+
+        # Calculate gas price
+        tx_PAYGAS_gasprice = computation.get_PAYGAS_gas_price()
+        if tx_PAYGAS_gasprice is None:
+            tx_PAYGAS_gasprice = 0
+        # Calculate gas usage
+        gas_remaining = computation.get_gas_remaining()
+        gas_refunded = computation.get_gas_refund()
+        gas_used = transaction.gas - gas_remaining
+        gas_refund = min(gas_refunded, gas_used // 2)
+        # Calculate transaction fee
+        tx_fee = (transaction.gas - gas_remaining - gas_refund) * tx_PAYGAS_gasprice
+        # Bookkeep this transaction fee
+        if hasattr(block, "tx_fee_sum"):
+            block.tx_fee_sum += tx_fee
```

Comment:
I'd like to figure out a way to do this which doesn't involve storing data on the block object.  Looking around at the code however, it's not very clear where that would be.

For now, lets just formalize this property on the block object so that we don't have to do `hasattr` checks.